ui/app/components/inference/ToolParametersSection.tsx

```diff
@@ -0,0 +1,270 @@
+                renderCard={(tool) => <ToolCard tool={tool} />}
+              />
+            </div>
+          )}
+
+          {/* Provider Tools */}
+          {hasProviderTools && (
+            <div className="flex flex-col gap-2">
+              <span className="text-fg-muted text-xs font-medium uppercase">
+                Provider Tools ({provider_tools.length})
+              </span>
+              <ToolsList
+                tools={provider_tools}
+                getLabel={getProviderToolLabel}
+                renderCard={(tool) => <ProviderToolCard tool={tool} />}
+              />
+            </div>
+          )}
+        </div>
+      </SnippetContent>
+    </SnippetLayout>
+  );
+}
+
+function ToolChoiceBadge({ toolChoice }: { toolChoice: ToolChoice }) {
+  if (typeof toolChoice === "string") {
+    return <Badge variant="outline">{toolChoice}</Badge>;
+  }
+
+  // toolChoice is { specific: string }
+  return (
+    <Badge variant="outline">
+      specific: <span className="font-mono">{toolChoice.specific}</span>
+    </Badge>
+  );
+}
+
+interface ToolsListProps<T> {
+  tools: T[];
+  getLabel: (tool: T, index: number) => string;
+  renderCard: (tool: T) => React.ReactNode;
+}
+
+function ToolsList<T>({ tools, getLabel, renderCard }: ToolsListProps<T>) {
+  if (tools.length === 1) {
+    return <>{renderCard(tools[0])}</>;
+  }
+
+  const tabs = tools.map((tool, index) => ({
+    id: String(index),
+    label: getLabel(tool, index),
+    content: renderCard(tool),
+  }));
+
+  return <SnippetTabs tabs={tabs} />;
+}
+
+function getToolName(tool: Tool): string {
+  if (tool.type === "function") {
+    return tool.name;
+  } else if (tool.type === "openai_custom") {
+    return tool.name;
+  }
+  return "Unknown Tool";
+}
+
+function ToolCard({ tool }: { tool: Tool }) {
+  if (tool.type === "function") {
+    return <FunctionToolCard tool={tool} />;
+  } else if (tool.type === "openai_custom") {
+    return <OpenAICustomToolCard tool={tool} />;
+  }
+  return null;
+}
+
+function FunctionToolCard({ tool }: { tool: FunctionTool }) {
+  return (
+    <div className="flex flex-col gap-2">
+      <div className="flex flex-wrap items-center gap-2">
+        <Badge variant="secondary">function</Badge>
+        <span className="font-mono text-sm font-medium">{tool.name}</span>
+        {tool.strict && (
+          <Badge variant="outline" className="text-xs">
+            strict
+          </Badge>
+        )}
+      </div>
+      {tool.description && (
+        <p className="text-fg-muted text-sm">{tool.description}</p>
+      )}
+      <div className="mt-1">
+        <span className="text-fg-muted mb-1 block text-xs font-medium uppercase">
+          Parameters
+        </span>
+        <CodeEditor
+          allowedLanguages={["json"]}
+          value={JSON.stringify(tool.parameters, null, 2)}
+          readOnly
+        />
+      </div>
+    </div>
+  );
+}
+
+function OpenAICustomToolCard({ tool }: { tool: OpenAICustomTool }) {
```

Comment:
`OpenAICustomToolCard` and `FunctionToolCard` are still pretty much the same, do you want to unify them too?